sympy/stats/rv.py

```diff
@@ -1076,57 +1081,73 @@ def sample(expr, condition=None, size=(), library='scipy',
+    >>> samp = sample(N)
+    >>> samp in N.pspace.domain.set
     True
-    >>> samp = next(sample(N, N>0)) # doctest: +SKIP
-    >>> samp > 0 # doctest: +SKIP
+    >>> samp = sample(N, N>0)
+    >>> samp > 0
     True
-    >>> samp_list = next(sample(N, size=4)) # doctest: +SKIP
-    >>> [sam in N.pspace.domain.set for sam in samp_list] # doctest: +SKIP
+    >>> samp_list = sample(N, size=4)
+    >>> [sam in N.pspace.domain.set for sam in samp_list]
     [True, True, True, True]
+    >>> sample(N, size = (2,3)) # doctest: +SKIP
+    array([[5.42519758, 6.40207856, 4.94991743],
+       [1.85819627, 6.83403519, 1.9412172 ]])
     >>> G = Geometric('G', 0.5) # Discrete Random Variable
-    >>> samp_list = next(sample(G, size=3)) # doctest: +SKIP
+    >>> samp_list = sample(G, size=3)
     >>> samp_list # doctest: +SKIP
-    array([10,  4,  1])
-    >>> [sam in G.pspace.domain.set for sam in samp_list] # doctest: +SKIP
+    [1, 3, 2]
+    >>> [sam in G.pspace.domain.set for sam in samp_list]
     [True, True, True]
     >>> MN = Normal("MN", [3, 4], [[2, 1], [1, 2]]) # Joint Random Variable
-    >>> samp_list = next(sample(MN, size=4)) # doctest: +SKIP
+    >>> samp_list = sample(MN, size=4)
     >>> samp_list # doctest: +SKIP
-    array([[4.22564264, 3.23364418],
-           [3.41002011, 4.60090908],
-           [3.76151866, 4.77617143],
-           [4.71440865, 2.65714157]])
-    >>> [tuple(sam) in MN.pspace.domain.set for sam in samp_list] # doctest: +SKIP
+    [array([2.85768055, 3.38954165]),
+     array([4.11163337, 4.3176591 ]),
+     array([0.79115232, 1.63232916]),
+     array([4.01747268, 3.96716083])]
+    >>> [tuple(sam) in MN.pspace.domain.set for sam in samp_list]
     [True, True, True, True]
 
+    .. versionchanged:: 1.7.0
+        sample returns an iterator containing the samples instead of value.
 
-    Returns
-    =======
-
-    sample: iterator object
-        iterator object containing the sample/samples of given expr
+    .. versionchanged:: 1.9.0
+        sample returns values or array of values instead of an iterator and numsamples is deprecated.
 
     """
-    ### TODO: Remove the user warnings in the future releases
-    message = ("The return type of sample has been changed to return an "
-                  "iterator object since version 1.7. For more information see "
-                  "https://github.com/sympy/sympy/issues/19061")
-    warnings.warn(filldedent(message))
-    return sample_iter(expr, condition, size=size, library=library,
+
+    iterator = sample_iter(expr, condition, size=size, library=library,
                                                         numsamples=numsamples, seed=seed)
 
+    if numsamples != 1:
+        SymPyDeprecationWarning(
+                 feature="numsamples parameter",
+                 issue=21563,
+                 deprecated_since_version="1.9").warn()
+
+        return [next(iterator) for i in range(numsamples)]
+
+    return next(iterator)
+
```

Comment:
What's the motivation behind this, though?

Reply:
If numsamples is 1, then we dont need to iterate over and a single float or array is returned. Otherwise, we need to iterate over to get all the samples in a list.
If we iterate over even when numsamples is 1, then it would return a list of length 1, which is unnecessary.
Also, since numsamples is deprecated, any value other than the default value of numsamples i.e 1 should cause a warning.

Reply:
Iteration is bad I believe. `shape`(`size` is confusing) is a better option. The first element can be the number of samples and the rest size of each sample. It would help in using the vectorization benefits provided by the underlying sampling library. But anyways keep it for a separate PR. I am fine with any choice.

Reply:
Yes, I agree shape is better. I will probably add that as a parameter in the next PR. I am trying to get the output right in this PR.

Reply:
Is the size parameter new in this PR?

If size was used previously then we should keep it for now for backwards compatibility.

Reply:
size parameter was there before the PR. We will keep it for backwards compatibility but we should move onto shape. It seems the best for sample.
Size and numsamples make sense for sample_iterator but not sample.